plugins/catalog-backend/src/ingestion/processors/LocationDiscoveryProcessor.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { UrlReader } from '@backstage/backend-common';
+import { LocationSpec } from '@backstage/catalog-model';
+import { Logger } from 'winston';
+import * as result from './results';
+import { CatalogProcessor, CatalogProcessorEmit } from './types';
+import parseGitUri, { stringify } from 'git-url-parse';
+
+type Options = {
+  reader: UrlReader;
+  logger: Logger;
+  pattern?: RegExp;
+};
+
+// TODO: This regex is wrong as ADR008 states its only "catalog-info.yaml"
+const defaultPattern = /catalog-info\.ya?ml$/;
+
+// TODO: This name is subject to change
+export class LocationDiscoveryProcessor implements CatalogProcessor {
+  constructor(private readonly options: Options) {}
+
+  async readLocation(
+    location: LocationSpec,
+    _: boolean,
+    emit: CatalogProcessorEmit,
+  ): Promise<boolean> {
+    if (location.type !== 'url') {
+      return false;
+    }
+
+    const pattern = this.options.pattern ?? defaultPattern;
+    const { filepath } = parseGitUri(location.target);
+
+    // TODO: Maybe find a better way to detect whether discovery is necessary
+    //  (e.g. by trying to download it, or by reading the tree first?)
+    if (pattern.test(filepath ?? '')) {
```

Comment:
Here we could also use `parseGitUri(target).filepathtype ? 'repo' : 'file'`

Reply:
As seen in https://github.com/backstage/backstage/pull/4014 - the filepathtype thing is not a particularly good check it seems, for some providers at least